Sources/Testing/Attachments/Test.Attachment.swift

```diff
@@ -0,0 +1,258 @@
+    _ attachableValue: borrowing some Test.Attachable & ~Copyable,
+    named preferredName: String? = nil,
+    sourceLocation: SourceLocation = #_sourceLocation
+  ) {
+    var proxyAttachable = _AttachableProxy()
+
+    // BUG: the borrow checker thinks that withErrorRecording() is consuming
+    // attachableValue, so get around it with an additional do/catch clause.
+    do {
+      let proxyAttachment = Self(proxyAttachable, named: preferredName, sourceLocation: sourceLocation)
+      proxyAttachable.encodedValue = try attachableValue.withUnsafeBufferPointer(for: proxyAttachment) { buffer in
+        [UInt8](buffer)
+      }
+    } catch {
+      Issue.withErrorRecording(at: sourceLocation) {
+        throw error
+      }
+    }
+
+    self.init(proxyAttachable, named: preferredName, sourceLocation: sourceLocation)
+  }
+}
+
+#if !SWT_NO_FILE_IO
+// MARK: - Writing
+
+extension Test.Attachment {
+  /// Write the attachment's contents to a file in the specified directory.
+  ///
+  /// - Parameters:
+  ///   - directoryPath: The directory that should contain the attachment when
+  ///     written.
+  ///
+  /// - Throws: Any error preventing writing the attachment.
+  ///
+  /// - Returns: The path to the file that was written.
+  ///
+  /// The attachment is written to a file _within_ `directoryPath`, whose name
+  /// is derived from the value of the ``Test/Attachment/preferredName``
+  /// property.
+  @_spi(ForToolsIntegrationOnly)
+  public func write(toFileInDirectoryAtPath directoryPath: String) throws -> String {
+    try write(
+      toFileInDirectoryAtPath: directoryPath,
+      appending: String(UInt64.random(in: 0 ..< .max), radix: 36)
+    )
+  }
+
+  /// Write the attachment's contents to a file in the specified directory.
+  ///
+  /// - Parameters:
+  ///   - directoryPath: The directory to which the attachment should be
+  ///     written.
+  ///   - suffix: A suffix to attach to the file name (instead of randomly
+  ///     generating one.) This value may be evaluated multiple times.
+  ///
+  /// - Throws: Any error preventing writing the attachment.
+  ///
+  /// - Returns: The path to the file that was written.
+  ///
+  /// The attachment is written to a file _within_ `directoryPath`, whose name
+  /// is derived from the value of the ``Test/Attachment/preferredName``
+  /// property and the value of `suffix`.
+  ///
+  /// If the argument `suffix` always produces the same string, the result of
+  /// this function is undefined.
+  func write(toFileInDirectoryAtPath directoryPath: String, appending suffix: @autoclosure () -> String) throws -> String {
+    let result: String
+
+    var file: FileHandle?
+    do {
+      // First, attempt to create the file with the exact preferred name. If a
+      // file exists at this path (note "x" in the mode string), an error will
+      // be thrown and we'll try again by adding a suffix.
+      let preferredPath = appendPathComponent(preferredName, to: directoryPath)
```

Comment:
Not sure if this is a minor bug with the appendPathComponent implementation, but if I pass "/tmp/" as the `directoryPath` on macOS the returned string ends up being `/tmp//myfile.txt` with a double slash.

Reply:
It's just because we don't have proper path parsing without Foundation—the double-slash is harmless. You'll note that setting `configuration.whateverINamedThatProperty` will `realpath()` it and remove the extra slash.